### PR TITLE
[GFX-2515] Merge material use flags to bitfield and merge intensities

### PR DESCRIPTION
**Pull request by gaborvalasek**
## Jira ticket URL (Why?)
[GFX-2515](https://shapr3d.atlassian.net/browse/GFX-2515)

## Short description (What? How?) 📖
We have reached the `MAX_PARAMETERS_COUNT = 48` cap in at least one of our material shaders in terms of  material parameters. It is not immediately obvious to me why this is capped at that exact value (UBOs have to offer at least 1024 locations - https://registry.khronos.org/OpenGL-Refpages/gl4/html/glGet.xhtml and GL_MAX_UNIFORM_LOCATIONS) but it's the same on upstream and I don't want to interfere with anything by altering the limit.

Since our material interface is pretty stable, we can finally address this properly by taking care of our material shader parameters. This PR does the following parameter count and size optimizations:
* all `use*` material attributes are now part of one big happy bitfield (`usageFlags`). This merges 15 different locations into 1
* the most common intensities (normal, clearcoat normal, specular, occlusion) are merged into a single `float4` attribute

In OpenGL lingo, this means that we moved 17 locations into 3, freeing up space for 14 parameters. 

## Upstreaming scope
This only affects our GltfViewer and material shaders, there's nothing to be upstreamed. 

### Design review 🎨
N/A

### Affected areas 🧭
Visualization materials

### Special use-cases to test 🧷
Make sure GltfViewer is still operating correctly (the artists are doing some of it already), if you toggle textures on and off, those changes are visible, etc.

### How did you test it? 🤔
Manual 💁‍♂️
By making a couple materials in GltfViewer, using our textures. Also verified that a couple of existing materials look the same as before.

Automated 💻
N/A (Shapr3D side integration will do the render tests)

[GFX-2515]: https://shapr3d.atlassian.net/browse/GFX-2515?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: filament/src/materials/MaterialHelpersShading.fxh [hunk 1]
```diff
@@ -12,6 +12,132 @@ float SignNoZero(float f) {
...
+// textures, how it should swizzle them, and/or alter lighting. These used to be separate uniform
+// integers but we simply ran out of the spare space and were forced to pack these together. 
+//
+// Old uniform int name	        Bit	    New query from flag bitfield
+// =====================================================================
+// useBaseColorTexture	        0	    materialParams.usageFlags & 1
+// useClearCoatNormalTexture	1	    materialParams.usageFlags & 2
+// useClearCoatRoughnessTexture	2	    materialParams.usageFlags & 4
+// useMetallicTexture	        3	    materialParams.usageFlags & 8
+// useNormalTexture	            4	    materialParams.usageFlags & 16
```
Comment (palver123) [inline filament/src/materials/MaterialHelpersShading.fxh]: nit: pls align the columns

File: filament/src/materials/MaterialHelpersShading.fxh [hunk 2]
```diff
@@ -12,6 +12,132 @@ float SignNoZero(float f) {
...
+// useOcclusionTexture	        5	    materialParams.usageFlags & 32
+// useRoughnessTexture	        6	    materialParams.usageFlags & 64
+// useSheenRoughnessTexture	    7	    materialParams.usageFlags & 128
+// useSwizzledNormalMaps	    8	    materialParams.usageFlags & 256
+// useThicknessTexture 	        9	    materialParams.usageFlags & 512
+// useTransmissionTexture 	    10	    materialParams.usageFlags & 1024
+// useWard	                    11	    materialParams.usageFlags & 2048
```
Comment (palver123) [inline filament/src/materials/MaterialHelpersShading.fxh]: `doDeriveXXX` are missing from the list

File: filament/src/materials/MaterialHelpersShading.fxh [hunk 3]
```diff
@@ -530,4 +656,4 @@ void ApplyAllPostPrepare(inout MaterialInputs material, inout FragmentData fragm
 
     ApplyNonTextured(material, fragmentData);
     ApplyShaprScalars(material, fragmentData);
-}
\ No newline at end of file
+}
```
Comment (palver123) [inline filament/src/materials/MaterialHelpersShading.fxh:662]: nit: missing newline

**Review by palver123 — APPROVED** [review]
The code looks OK to me. Did you measure any perf. boost on the GPU-side in Shapr3D?

**Comment by gaborvalasek** [comment]
> The code looks OK to me. Did you measure any perf. boost on the GPU-side in Shapr3D?

I haven't tried it but, unfortunately, this is still less than a cacheline worth of savings in memory bandwidth, so any gains we have is most likely buried beneath the measurement noise.

File: filament/src/materials/MaterialHelpersShading.fxh (hunk 1 above)
Comment (gaborvalasek) [reply filament/src/materials/MaterialHelpersShading.fxh]: Fixed in c680e4f2bf6838ae0f7480a18558640779362379

File: filament/src/materials/MaterialHelpersShading.fxh (hunk 2 above)
Comment (gaborvalasek) [reply filament/src/materials/MaterialHelpersShading.fxh]: Fixed in c680e4f2bf6838ae0f7480a18558640779362379

File: filament/src/materials/MaterialHelpersShading.fxh (hunk 3 above)
Comment (gaborvalasek) [reply filament/src/materials/MaterialHelpersShading.fxh:662]: Weird, locally, I have a newline but I've added an extra one in c680e4f2bf6838ae0f7480a18558640779362379

Comment (palver123) [reply filament/src/materials/MaterialHelpersShading.fxh:662]: I think it depends on the editor you open it with. I trust HxD on Windows, or Hex Fiend on macOS 🔡

Comment (veghbalazs) [reply filament/src/materials/MaterialHelpersShading.fxh:662]: It's the opposite, @gaborvalasek's changes in 9959d5a3cb7adc0309d68b951171fb8858ee894c added the newline which had been missing from master. (Remember, this is how GitHub shows this, the red icons signals **missing** newline).
image

**Pull request merged**